torch/csrc/utils/python_strings.h

```diff
@@ -35,6 +35,31 @@ inline std::string THPUtils_unpackString(PyObject* obj) {
   throw std::runtime_error("unpackString: expected bytes or unicode object");
 }
 
+// Unpacks PyBytes (PyString) or PyUnicode as c10::string_view
+// PyBytes are unpacked as-is. PyUnicode is unpacked as UTF-8.
+// NOTE: If `obj` is destroyed, then the non-owning c10::string_view will
+//   become invalid. If the string needs to be accessed at any point after
+//   `obj` is destroyed, then the c10::string_view should be copied into
+//   a std::string, or another owning object, and kept alive. For an example,
+//   look at how IValue and autograd nodes handle c10::string_view arguments.
+// NOTE: this method requires the GIL
+inline c10::string_view THPUtils_unpackStringView(PyObject* obj) {
+  if (PyBytes_Check(obj)) {
+    size_t size = PyBytes_GET_SIZE(obj);
+    return c10::string_view(PyBytes_AS_STRING(obj), size);
+  }
+  if (PyUnicode_Check(obj)) {
+    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+    Py_ssize_t size;
+    const char* data = PyUnicode_AsUTF8AndSize(obj, &size);
```

Comment:
was reading the CPython docs for this and apparently they need to decode the unicode object and then cache the utf-8 rep. Wild.

Reply:
> was reading the CPython docs for this and apparently they need to decode the unicode object and then cache the utf-8 rep. Wild.

yes, but it's worth noting that since version 3.3, CPython uses an internal representation which contains an optimized special case for ASCII-only strings. In that case, the utf8 internal cache corresponds to the raw data so there is no extra cost in calling `PyUnicode_AsUTF8AndSize`. See [PEP 393](https://www.python.org/dev/peps/pep-0393/) for details.